### PR TITLE
dpdk: add performance test using dpdk tool

### PR DESCRIPTION
**Pull request by quanwenli**
ID:1378

Signed-off-by: Wenli Quan <wquan@redhat.com>

### CODE REVIEW

**Comment by quanwenli** [comment]
JOB LOG    : /root/avocado/job-results/job-2023-09-19T09.56-8ac2be4/job.log
 (1/1) Host_RHEL.m9.u4.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.vdpa_dpdk.q35: STARTED
 (1/1) Host_RHEL.m9.u4.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.vdpa_dpdk.q35: PASS (317.34 s)

**Comment by quanwenli** [comment]
@PaulYuuu , could you help review? thanks

File: provider/dpdk_utils.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
+import subprocess
+import paramiko
```
Comment (PaulYuuu) [inline provider/dpdk_utils.py:2]: Maybe it can be removed? You import it in line 9

File: provider/dpdk_utils.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+# Ensure paramiko is installed
+for pip in ['pip3', 'pip']:
+    try:
+        subprocess.check_call([pip, 'install', '--default-timeout=100', 'paramiko'])
+        import paramiko
+        break
+    except ImportError:
+        continue
+else:
+    print("Failed to install paramiko. Please install it manually.")
```
Comment (PaulYuuu) [inline provider/dpdk_utils.py:14]: For test log, we can use LOG_JOB
```python
LOG_JOB = logging.getLogger('avocado.test')
LOG_JOB.info(xxx)
```

Or would you like it to print to the terminal?

File: provider/dpdk_utils.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    if session.cmd_status("lsmod | grep -wq vfio"):
+        session.cmd("modprobe -r vfio-pci")
+        print("vfio-pci module removed")
+
+    if session.cmd_status("lsmod | grep -wq vfio_pci"):
+        session.cmd("modprobe -r vfio")
+        print("vfio module removed")
+
+    session.cmd_output("modprobe vfio enable_unsafe_noiommu_mode=Y")
+    print("vfio module loaded")
```
Comment (PaulYuuu) [inline provider/dpdk_utils.py:45]: Same above.

File: qemu/tests/cfg/vdpa_dpdk.cfg [hunk 4]
```diff
@@ -0,0 +1,42 @@
+- vdpa_dpdk:
+    no JeOS
+    no Windows
+    no Host_RHEL.m5, Host_RHEL.m6, Host_RHEL.m7
+    virt_test_type = qemu
+    type = vdpa_dpdk
+    kill_vm = yes
+    machine_type = q35
```
Comment (PaulYuuu) [inline qemu/tests/cfg/vdpa_dpdk.cfg:8]: `only q35`?

File: provider/dpdk_utils.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
+import subprocess
```
Comment (PaulYuuu) [inline provider/dpdk_utils.py:1]:
```suggestion
from avocado.utils import process
```

File: provider/dpdk_utils.py (hunk 2 above)
Comment (quanwenli) [reply provider/dpdk_utils.py:14]: For the vDPA feature, some cases need to be run on the host. therefore, it's preferable for the dpdk_utils.py file to be usable independently, making it unsuitable to use LOG_JOB.

File: provider/dpdk_utils.py (hunk 3 above)
Comment (quanwenli) [reply provider/dpdk_utils.py:45]: For the vDPA feature, some cases need to be run on the host. therefore, it's preferable for the dpdk_utils.py file to be usable independently, making it unsuitable to use LOG_JOB.

File: qemu/tests/cfg/vdpa_dpdk.cfg (hunk 4 above)
Comment (quanwenli) [reply qemu/tests/cfg/vdpa_dpdk.cfg:8]: you are right, I will remove it .

File: provider/dpdk_utils.py (hunk 1 above)
Comment (quanwenli) [reply provider/dpdk_utils.py:2]: I wil remove it.

File: provider/dpdk_utils.py (hunk 5 above)
Comment (quanwenli) [reply provider/dpdk_utils.py:1]: For the vDPA feature, some cases need to be run on the host. therefore, it's preferable for the dpdk_utils.py file to be usable independently, making it unsuitable to use avocado.utils.process.

I have updated, please help review. thanks.

File: provider/dpdk_utils.py (hunk 2 above)
Comment (PaulYuuu) [reply provider/dpdk_utils.py:13]: OK, thank you for your explanation.

File: qemu/tests/cfg/vdpa_dpdk.cfg [hunk 6]
```diff
@@ -0,0 +1,41 @@
+- vdpa_dpdk:
+    no JeOS
+    no Windows
```
Comment (PaulYuuu) [inline qemu/tests/cfg/vdpa_dpdk.cfg:3]: Seems missing `only q35` here.

File: provider/dpdk_utils.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+# Ensure paramiko is installed
+for pip in ['pip3', 'pip']:
+    try:
+        subprocess.check_call([pip, 'install', '--default-timeout=100', 'paramiko'])
+        import paramiko
+        break
+    except ImportError:
+        continue
+else:
+    print("Failed to install paramiko. Please install it manually.")
```
[the remark this answers is not in the document]
Comment (quanwenli) [reply provider/dpdk_utils.py:14]: yes, I have update the code.

File: qemu/tests/cfg/vdpa_dpdk.cfg (hunk 6 above)
Comment (quanwenli) [reply qemu/tests/cfg/vdpa_dpdk.cfg:3]: I have confirmed with developer , vdpa support with/without q35, so I remove the q35.

Comment (PaulYuuu) [reply qemu/tests/cfg/vdpa_dpdk.cfg:3]: OK, as RHEL only x86 supports vdpa, but other users can rebuild kernel to support it, so let's keep your current code.
Could you provide latest result?

**Comment by quanwenli** [comment]
JOB LOG    : /root/avocado/job-results/job-2023-09-26T03.03-55f4622/job.log
 (1/1) Host_RHEL.m9.u4.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.vdpa_dpdk.q35: STARTED
 (1/1) Host_RHEL.m9.u4.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.vdpa_dpdk.q35: PASS (302.77 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2023-09-26T03.03-55f4622/results.html
JOB TIME   : 304.36 s

**Review by PaulYuuu — APPROVED** [review]

**Pull request merged**